catalog/core/catalog-core-directorymonitor/src/main/java/org/codice/ddf/catalog/content/monitor/AsyncFileAlterationObserver.java

```diff
@@ -0,0 +1,362 @@
+   */
+  private void commitDelete(AsyncFileEntry entry, boolean success) {
+    LOGGER.debug("commitDelete({},{}): Starting...", entry.getName(), success);
+    if (success) {
+      entry.getParent().ifPresent(e -> e.removeChild(entry));
+      entry.destroy();
+    }
+    removeFromProcessors(entry);
+  }
+
+  /**
+   * Steps file by file comparing the snapshot state to the current state of the directory being
+   * monitored.
+   *
+   * @param parent The parent directory (Wrapped in a AsyncFileEntry)
+   * @param previous The list of all children of the parent directory (In sorted order)
+   * @param files The list of current files (in sorted order)
+   */
+  private void checkAndNotify(
+          final AsyncFileEntry parent, final List<AsyncFileEntry> previous, final File[] files) {
+    int c = 0;
+    for (final AsyncFileEntry entry : previous) {
+      while (c < files.length && entry.compareToFile(files[c]) > 0) {
+        doCreate(new AsyncFileEntry(parent, files[c]));
+        c++;
+      }
+      if (c < files.length && entry.compareToFile(files[c]) == 0) {
+        //  Has the file been committed before? We may have matched it
+        //  but if it doesn't exist downstream then a create would be appropriate
+        if (entry.isInitCommit()) {
+          doMatch(entry);
+          checkAndNotify(entry, entry.getChildren(), listFiles(files[c]));
+        } else {
+          doCreate(entry);
+        }
+        c++;
+      } else {
+        //  Do Delete
+        if (!entry.checkNetwork()) {
+          //  The file MAY still exist but it's the network that's down.
+          return;
+        }
+        checkAndNotify(entry, entry.getChildren(), FileUtils.EMPTY_FILE_ARRAY);
+        doDelete(entry);
+      }
+    }
+    for (; c < files.length; c++) {
+      doCreate(new AsyncFileEntry(parent, files[c]));
+    }
+  }
+
+  /**
+   * Called when the observer should compare the snapshot state to the actual state of the directory
+   * being monitored.
+   */
+  public void checkAndNotify() {
+
+    if (listener == null) {
+      return;
+    }
+
+    /* fire onStart() */
+    listener.onStart(this);
+
+    /* fire directory/file events */
+    final File root = rootFile.getFile();
+    if (root.exists()) {
+      checkAndNotify(rootFile, rootFile.getChildren(), listFiles(rootFile.getFile()));
+    } else if (rootFile.exists()) {
+      checkAndNotify(rootFile, rootFile.getChildren(), FileUtils.EMPTY_FILE_ARRAY);
+    } else {
+      // Didn't exist and still doesn't
+    } //  rootFile.refresh()
+
+    /* fire onStop() */
+    listener.onStop(this);
+  }
+
+  /**
+   * listFiles:
+   *
+   * <p>Note: returns a new Array to avoid sync access exceptions
+   *
+   * @return A new sorted File Array
+   */
+  private File[] listFiles(File file) {
+    if (file.isDirectory()) {
+      File[] temp = file.listFiles();
+      Arrays.sort(temp);
+      return temp;
+    }
+    return FileUtils.EMPTY_FILE_ARRAY;
+  }
+
+  /**
```

Comment:
📘 Kept all the processing queue together. I didn't know if I wanted to make it it's own file or if these sync blocks are required for a concurrentSkipListSet but I wasn't sure so I kept them. If they aren't then these methods are not needed